### PR TITLE
pageserver: add GcError type

### PR DESCRIPTION
**Pull request by jcsp**
## Problem

- Because GC exposes all errors as an anyhow::Error, we have intermittent issues with spurious log errors during shutdown, e.g. in this failure of a performance test https://neon-github-public-dev.s3.amazonaws.com/reports/main/9300804302/index.html#suites/07874de07c4a1c9effe0d92da7755ebf/214a2154f6f0217a/

```
Gc failed 1 times, retrying in 2s: shutting down
```

GC really doesn't do a lot of complicated IO: it doesn't benefit from the backtrace capabilities of anyhow::Error, and can be expressed more robustly as an enum.

## Summary of changes

- Add GcError type and use it instead of anyhow::Error in GC functions
- In `gc_iteration_internal`, return GcError::Cancelled on shutdown rather than Ok(()) (we only used Ok before because we didn't have a clear cancellation error variant to use).
- In `gc_iteration_internal`, skip past timelines that are shutting down, to avoid having to go through another GC iteration if we happen to see a deleting timeline during a GC run.
- In `refresh_gc_info_internal`, avoid an error case where a timeline might not be found after being looked up, by carrying an Arc<Timeline> instead of a TimelineId between the first loop and second loop in the function.
- In HTTP request handler, handle Cancelled variants as 503 instead of turning all GC errors into 500s.

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: pageserver/src/tenant.rs [hunk 1]
```diff
@@ -487,6 +487,33 @@ enum CreateTimelineCause {
...
+    #[error("not active")]
+    NotActive,
+
+    // A requested GC cutoff LSN was invalid, for example it tried to move backwards
+    #[error("not active")]
+    BadLsn { why: String },
+
+    // A remote storage error while scheduling updates after compaction
+    #[error(transparent)]
+    RemoteUpload(anyhow::Error),
```
Comment (arpad-m) [inline pageserver/src/tenant.rs:510]: we also delete stuff, right? So maybe `RemoteOperation` is a better name?

File: pageserver/src/tenant.rs [hunk 2]
```diff
@@ -487,6 +487,33 @@ enum CreateTimelineCause {
     Delete,
 }
 
+#[derive(thiserror::Error, Debug)]
+pub(crate) enum GcError {
+    // The tenant is shutting down
+    #[error("tenant shutting down")]
+    TenantCancelled,
```
Comment (arpad-m) [inline pageserver/src/tenant.rs:494]: Do we need to distinguish the two cases? Cancellation is cancellation. Also, the names are probably not accurate because at least one of the two gets invoked during pageserver shutdowns as well.

File: pageserver/src/tenant.rs [hunk 3]
```diff
@@ -487,6 +487,33 @@ enum CreateTimelineCause {
...
+    #[error("not active")]
+    BadLsn { why: String },
+
+    // A remote storage error while scheduling updates after compaction
+    #[error(transparent)]
+    RemoteUpload(anyhow::Error),
+
+    // If GC was invoked for a particular timeline, this error means it didn't exist
+    #[error("timeline not found")]
+    NotFound,
```
Comment (arpad-m) [inline pageserver/src/tenant.rs:514]:
```suggestion
    TimelineNotFound,
```
if the *tenant* was not found we don't yield this error, right? because then from how I understand it, already the http handler errors. the rustdoc and error message describe this well but the name is a bit generic and easy to misunderstand.

File: pageserver/src/tenant.rs (hunk 2 above)
Comment (jcsp) [reply pageserver/src/tenant.rs:494]: I started out with just one Cancelled variant.  Then, I was updating the code around this, and it felt quite error prone in the cases where we want to drop out of GC when the tenant is cancelled, but not with just one timeline is cancelled.

TLDR: we could make do with just one Cancelled, but I think it would make the code less obvious in the places we handle it.

File: pageserver/src/tenant.rs (hunk 1 above)
Comment (jcsp) [reply pageserver/src/tenant.rs]: Yeah, the Upload part was redundant anyway so I just shortened this to `Remote`.  950e5d235

File: pageserver/src/tenant.rs (hunk 3 above)
Comment (jcsp) [reply pageserver/src/tenant.rs]: Right, when we call into GC it's always implicitly true that the tenant exists.  Changed in 950e5d235

**Review by arpad-m — APPROVED** [review]

File: pageserver/src/tenant.rs (hunk 2 above)
Comment (arpad-m) [reply pageserver/src/tenant.rs:494]: Hmmm I see that explanation makes sense, approving
